blockbuster/abci/abci.go

```diff
@@ -18,70 +15,36 @@ type (
 		logger              log.Logger
 		mempool             blockbuster.Mempool
 		txEncoder           sdk.TxEncoder
+		prepareLanesHandler blockbuster.PrepareLanesHandler
 		processLanesHandler blockbuster.ProcessLanesHandler
 	}
 )
 
-// NewProposalHandler returns a new ProposalHandler.
+// NewProposalHandler returns a new abci++ proposal handler.
 func NewProposalHandler(logger log.Logger, mempool blockbuster.Mempool, txEncoder sdk.TxEncoder) *ProposalHandler {
 	return &ProposalHandler{
 		logger:              logger,
 		mempool:             mempool,
 		txEncoder:           txEncoder,
+		prepareLanesHandler: ChainPrepareLanes(mempool.Registry()...),
 		processLanesHandler: ChainProcessLanes(mempool.Registry()...),
 	}
 }
 
-// ChainProcessLane chains together the proposal verification logic from each lane
-// into a single function. The first lane in the chain is the first lane to be verified and
-// the last lane in the chain is the last lane to be verified.
-func ChainProcessLanes(chain ...blockbuster.Lane) blockbuster.ProcessLanesHandler {
-	if len(chain) == 0 {
-		return nil
-	}
-
-	// Handle non-terminated decorators chain
-	if (chain[len(chain)-1] != terminator.Terminator{}) {
-		chain = append(chain, terminator.Terminator{})
-	}
-
-	return func(ctx sdk.Context, proposalTxs [][]byte) (sdk.Context, error) {
-		return chain[0].ProcessLane(ctx, proposalTxs, ChainProcessLanes(chain[1:]...))
-	}
-}
-
+// PrepareProposalHandler prepares the proposal by selecting transactions from each lane
+// according to each lane's selection logic. We select transactions in a greedy fashion. Note that
+// each lane has an boundary on the number of bytes that can be included in the proposal. By default,
+// the default lane will not have a boundary on the number of bytes that can be included in the proposal and
+// will include all valid transactions in the proposal (up to MaxTxBytes).
 func (h *ProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHandler {
 	return func(ctx sdk.Context, req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-		var (
-			selectedTxs  = make(map[string][]byte)
-			totalTxBytes int64
-		)
-
-		for _, l := range h.mempool.Registry() {
-			if totalTxBytes < req.MaxTxBytes {
-				laneTxs, err := l.PrepareLane(ctx, req.MaxTxBytes, selectedTxs)
-				if err != nil {
-					h.logger.Error("failed to prepare lane; skipping", "lane", l.Name(), "err", err)
-					continue
-				}
-
-				for _, txBz := range laneTxs {
-					totalTxBytes += int64(len(txBz))
-
-					txHash := sha256.Sum256(txBz)
-					txHashStr := hex.EncodeToString(txHash[:])
+		proposal := h.prepareLanesHandler(ctx, blockbuster.Proposal{
```

Comment:
Is the idea that `prepareLanesHandler` modifies the `blockbuster.Proposal` argument passed in? If so, let's make sure we pass in a pointer.